spec/abstract/Decode_A.thy

```diff
@@ -515,47 +509,62 @@ where
         odE
       | _ \<Rightarrow> throwError (InvalidCapability 1);
       returnOk $ InvokeSchedContextBind sc_ptr cap
-    odE
-  | SchedContextUnbindObject \<Rightarrow> doE
+    odE"
+
+definition
+  decode_sched_context_unbind_object ::
+  "obj_ref \<Rightarrow> cap list \<Rightarrow> (sched_context_invocation, 'z::state_ext) se_monad"
+where
+  "decode_sched_context_unbind_object sc_ptr excaps \<equiv> doE
       whenE (length excaps = 0) $ throwError TruncatedMessage;
       cap \<leftarrow> returnOk $ hd excaps;
       case cap of
         ThreadCap tcb_ptr \<Rightarrow> doE
-          sc_tcb_opt \<leftarrow> liftE $ get_sc_obj_ref sc_tcb sc_ptr;
-          whenE (sc_tcb_opt \<noteq> Some tcb_ptr) $ throwError IllegalOperation;
+          sc \<leftarrow> liftE $ get_sched_context sc_ptr;
+          whenE (sc_tcb sc \<noteq> Some tcb_ptr) $ throwError IllegalOperation;
           ct_ptr \<leftarrow> liftE $ gets cur_thread;
           whenE (tcb_ptr = ct_ptr) $ throwError IllegalOperation
         odE
       | NotificationCap ntfn_ptr _ _ \<Rightarrow> doE
-          sc_ntfn_opt \<leftarrow> liftE $ get_sc_obj_ref sc_ntfn sc_ptr;
-          whenE (sc_ntfn_opt \<noteq> Some ntfn_ptr) $ throwError IllegalOperation
+          sc \<leftarrow> liftE $ get_sched_context sc_ptr;
+          whenE (sc_ntfn sc \<noteq> Some ntfn_ptr) $ throwError IllegalOperation
         odE
       | _ \<Rightarrow> throwError (InvalidCapability 1);
       returnOk $ InvokeSchedContextUnbindObject sc_ptr cap
-    odE
-  | SchedContextUnbind \<Rightarrow> doE
-      cap \<leftarrow> returnOk $ hd excaps;
-      tcb_ptr \<leftarrow> returnOk $ obj_ref_of cap;
-      ct_ptr \<leftarrow> liftE $ gets cur_thread;
-      whenE (tcb_ptr = ct_ptr) $ throwError IllegalOperation;
-      returnOk $ InvokeSchedContextUnbind sc_ptr cap odE
-  | SchedContextYieldTo \<Rightarrow> doE
+    odE"
+
+definition
+  decode_sched_context_yield_to ::
+  "obj_ref \<Rightarrow> obj_ref option \<Rightarrow> (sched_context_invocation, 'z::state_ext) se_monad"
+where
+  "decode_sched_context_yield_to sc_ptr buffer \<equiv> doE
       sc \<leftarrow> liftE $ get_sched_context sc_ptr;
-      case sc_tcb sc of
-        None \<Rightarrow> throwError IllegalOperation
-      | Some tcb_ptr \<Rightarrow> doE
-          ct_ptr \<leftarrow> liftE $ gets cur_thread;
-          whenE (tcb_ptr = ct_ptr) $ throwError IllegalOperation;
-          prios \<leftarrow> liftE $ thread_get tcb_priority tcb_ptr;
-          ct_mcp \<leftarrow> liftE $ thread_get tcb_mcpriority ct_ptr;
-          whenE (prios > ct_mcp) $ throwError IllegalOperation
-      odE;
-\<^cancel>\<open>     whenE (sc_yield_from sc \<noteq> None) $ throwError IllegalOperation;\<close>
       ct_ptr \<leftarrow> liftE $ gets cur_thread;
+      whenE (the (sc_tcb sc) = ct_ptr) $ throwError IllegalOperation;
+      whenE (sc_tcb sc = None) $ throwError IllegalOperation;
```

Comment:
It's a bit confusing but I think that we want to either reverse these lines so that we know that we can safely use `the (sc_tcb sc)`, or change the first check to `sc_tcb sc = Some ct_ptr`. Otherwise it's possible to get some unintentional behaviour with `undefined`.

My guess is that as it is currently we might run in to problems with this in crefine.

Reply:
If I'm correct in wanting to make this change, we also need to update the Haskell.